openapi3/internalize_refs.go

```diff
@@ -0,0 +1,373 @@
+			r.Ref = "#/components/requestBodies/" + name
+			return r.Ref
+		}
+		if spec.Components.RequestBodies == nil {
+			spec.Components.RequestBodies = make(RequestBodies)
+		}
+		spec.Components.RequestBodies[name] = &RequestBodyRef{Value: r.Value}
+		r.Ref = "#/components/requestBodies/" + name
+		return r.Ref
+	}
+
+	addResponseToSpec := func(r *ResponseRef) string {
+		if r == nil || !isExternalRef(r.Ref) {
+			return ""
+		}
+		name := refNameResolver(r.Ref)
+		if _, ok := spec.Components.Responses[name]; ok {
+			r.Ref = "#/components/responses/" + name
+			return r.Ref
+		}
+		if spec.Components.Responses == nil {
+			spec.Components.Responses = make(Responses)
+		}
+		spec.Components.Responses[name] = &ResponseRef{Value: r.Value}
+		r.Ref = "#/components/responses/" + name
+		return r.Ref
+	}
+
+	addSecuritySchemeToSpec := func(ss *SecuritySchemeRef) string {
+		if ss == nil || !isExternalRef(ss.Ref) {
+			return ""
+		}
+		name := refNameResolver(ss.Ref)
+		if _, ok := spec.Components.SecuritySchemes[name]; ok {
+			ss.Ref = "#/components/securitySchemes/" + name
+			return ss.Ref
+		}
+		if spec.Components.SecuritySchemes == nil {
+			spec.Components.SecuritySchemes = make(SecuritySchemes)
+		}
+		spec.Components.SecuritySchemes[name] = &SecuritySchemeRef{Value: ss.Value}
+		ss.Ref = "#/components/securitySchemes/" + name
+		return ss.Ref
+	}
+
+	addExampleToSpec := func(e *ExampleRef) string {
+		if e == nil || !isExternalRef(e.Ref) {
+			return ""
+		}
+		name := refNameResolver(e.Ref)
+		if _, ok := spec.Components.Examples[name]; ok {
+			e.Ref = "#/components/examples/" + name
+			return e.Ref
+		}
+		if spec.Components.Examples == nil {
+			spec.Components.Examples = make(Examples)
+		}
+		spec.Components.Examples[name] = &ExampleRef{Value: e.Value}
+		e.Ref = "#/components/examples/" + name
+		return e.Ref
+	}
+
+	addLinkToSpec := func(l *LinkRef) string {
+		if l == nil || !isExternalRef(l.Ref) {
+			return ""
+		}
+		name := refNameResolver(l.Ref)
+		if _, ok := spec.Components.Links[name]; ok {
+			l.Ref = "#/components/links/" + name
+			return l.Ref
+		}
+		if spec.Components.Links == nil {
+			spec.Components.Links = make(Links)
+		}
+		spec.Components.Links[name] = &LinkRef{Value: l.Value}
+		l.Ref = "#/components/links/" + name
+		return l.Ref
+	}
+
+	addCallbackToSpec := func(c *CallbackRef) string {
+		if c == nil || !isExternalRef(c.Ref) {
+			return ""
+		}
+		name := refNameResolver(c.Ref)
+		if _, ok := spec.Components.Callbacks[name]; ok {
+			c.Ref = "#/components/callbacks/" + name
+			return c.Ref
+		}
+		if spec.Components.Callbacks == nil {
+			spec.Components.Callbacks = make(Callbacks)
+		}
+		spec.Components.Callbacks[name] = &CallbackRef{Value: c.Value}
+		c.Ref = "#/components/callbacks/" + name
+		return c.Ref
+	}
+
+	var derefSchema func(*Schema, []*Schema)
+	derefSchema = func(s *Schema, stack []*Schema) {
+		if len(stack) > 1000 {
+			panic("unresolved circular reference")
```

Comment:
Return an error here instead.
Do you have an example triggering this clause? If you do I'd like a test for this, if not isn't this dead code?

Reply:
Yeah I don't, I can remove it